docs/development/CODE_REVIEW_GUIDELINES.md

```diff
@@ -0,0 +1,121 @@
+
+### Expectations
+
+**Pull Request Authors:**
+The author is responsible for tracking up-to-date next actions for a Pull Request to progress towards being merged.
+
+**Reviewers:**
+Reviewers should prefer engaging in code review over starting new work (i.e. taking planned work items that haven't been started yet).
+
+Reviewers (and prospective reviewers) are encouraged to engage in reviews of codebases outside the projects and technologies they use on a day-to-day basis (but not expected to provide an approving review).
+
+## Best Practices
+
+### Reviewing
+
+### Starting a Review
+
+Consider if it could be broken into smaller Pull Requests. If it is clear that it can be, summarize your thinking on how in your Review.
+
+If the commits be (re)organized (i.e. reordered and/or amended) such that there is a commit at which the tests are passing prior to the conclusion of the main change, that's a signal that there's likely a logic split which can be made at that point in such a (re)arrangement.
+
+Use the following guidelines to evaluate whether a Pull Request should be approved:
+
+- *Every* Pull Request should have tests or justification otherwise; esp. bug fixes.
+- *Every* Pull Request should have at least 1 approval from a team member internal or external to the project. Exceptions made by repository maintainer(s), as necessary, on a case-by-case basis.
+
+During review, submit feedback using line comments (A); prefer "Add\[ing a\] single comment" over "Start[ing] a review" (B).
+
+*(Once a review has been started, the option to add single comments is removed. Preferring single comments allows feedback to happen even in the event of an interrupted review.)*
+
+### A
+
+![github_line_comment.png](./assets/github_line_comment.png)
+
+### B
+
+![line_comment_dialog_start.png](./assets/line_comment_dialog_start.png)
+
+**Referencing Issues Across Repositories:**
+When referencing issues from one repository, in another's Issues and Pull Requests, GitHub supports automatic links in markdown using the following format: `<org name>/<repo name>#<issue or PR number>`.`
+
+### Finishing a Review
+
+Write a summary of your observations and consider including positive remarks in addition to any constructive criticism.
+
+If you observe a deviation from these practices or another reason that this change should not be merged, select "request changes" and include a summary of the observation(s), as well as any practice(s) you find them to be in conflict with, in the review body (C).
+
+If you don't feel comfortable giving approval or requesting changes but want to share a summary or observations of larger patterns in the codebase or the company, select "Comment" and submit your review (C).
+
+Confirm that all items in the "required checklist" are checked or not applicable.
+
+If you believe the Pull Request looks good to merge, select "Approve" and submit your review (C).
+
+### C
+
+![submit_review.png](./assets/submit_review.png)
+
+### Merging
```

Comment:
Also, when merging, some rebasing / force pushing may be necessary: e.g. when CHANGELOG files have been modified by other merged PRs, which will block merge commits due to the commit hooks failure, and leave rebase as the only option (unless I am overlooking some git feature). Some hints regarding such scenarios might be helpful.

Reply:
Because we "squash & merge" our PRs to main (since some PRs might end up having dozens of commits), the simplest thing to do is to simply do a `git merge main` rather than a `git rebase main`. It saves times and avoids needing to force push.

Reply:
> I suggest we add a section: `Addressing Reviews` to help with a few items, e.g.:
> 
> * Do not rebase (and subsequently push -f) the PR branch, as it would make it challenging for the reviewer to identify what has changed between their review and the latest PR update. Instead push more commits to address review comments. The squash merge method will combine all the PR commits into one.

I have a strong opinion about **supporting rebase workflows** to whatever extent possible as I believe it can add significant value by improving readability of PRs on multiple levels. I've been developing and using a workflow I've pet-named "quantum git theory" which preserves the sequence of events on the PR timeline with respect to human interactions / "observations" (the git-time continuum). 

TLDR; contributors can rewrite history **up to the most recent human interaction** (e.g.: code comment, review, commit). If done correctly, this has the effect of **preserving the order of human-backed PR timeline events** with respect to commits (which I think is the main concern).

As @Olshansk points out, the consequences of what we do in version control on any given PR are constrained to that context of that branch due to the squash merge with main.

Perhaps we can discuss it further in a protocol hour. I have some other git-related things that might be useful to share with the team; perhaps, vice-versa.

Reply:
> Some hints regarding such scenarios might be helpful.

Agreed! :raised_hands: 

> Also, when merging, some rebasing / force pushing may be necessary: e.g. when CHANGELOG files have been modified by other merged PRs, which will block merge commits due to the commit hooks failure, and leave rebase as the only option (unless I am overlooking some git feature). 

AFAIK merging should never necessitate force pushing on its own. When merging an upstream branch which contains a conflict you can either resolve the conflict *in the merge commit* OR "accept theirs" (remote) and re-apply your changes in a subsequent commit. In either case, unless you changed history prior to the merge, it should be a fast-forward from the remote HEAD to your local HEAD (post-merge, etc.).

Here's an excerpt from #461 > "testing methodology" > "merge conflict test" which  illustrates the simplest variation of scenario:

  ```mermaid
  gitGraph
      commit
      branch FeatureA
      commit id: "[v0.0.0.1 - 2023-04-03]" type: HIGHLIGHT
      checkout main
      commit id: "[v0.0.0.1 - 2023-04-01]" type: HIGHLIGHT
      commit id: "[v0.0.0.2 - 2023-04-05]" type: HIGHLIGHT
      checkout FeatureA
      merge main
  ```